File: matlab-sim/ofdm_modulate.m
% FUNCTION: ofdm_modulate()
% This function performs OFDM modulation before data transmission.
function signal_tx = ofdm_modulate(data_tx, ifft_size, carriers, conj_carriers, carrier_count, symb_size, guard_time, fig)
    % Symbols per carrier for this frame
    carrier_symb_count = ceil(length(data_tx) / carrier_count);

    % Append zeros to data with a length not multiple of the number of carriers
    if length(data_tx) / carrier_count ~= carrier_symb_count
        padding = zeros(1, carrier_symb_count * carrier_count);
        padding(1:length(data_tx)) = data_tx;
        data_tx = padding;
    end

    % Serial to parallel: each column represents a carrier
    data_tx_matrix = reshape(data_tx, carrier_count, carrier_symb_count)';

    % Differential Encoding
    carrier_symb_count = size(data_tx_matrix, 1) + 1;
    diff_ref = round(rand(1, carrier_count) * (2^symb_size) + 0.5);
    data_tx_matrix = [diff_ref; data_tx_matrix];
    for k = 2:size(data_tx_matrix, 1)
        data_tx_matrix(k, :) = rem(data_tx_matrix(k, :) + data_tx_matrix(k-1, :), 2^symb_size);
    end

    % PSK (Phase Shift Keying) modulation
    [X, Y] = pol2cart(data_tx_matrix * (2*pi / (2^symb_size)), ones(size(data_tx_matrix)));
    complex_matrix = X + 1i * Y;

    % Assign IFFT bins to carriers and imaged carriers
    spectrum_tx = zeros(carrier_symb_count, ifft_size);
    spectrum_tx(:, carriers) = complex_matrix;
    spectrum_tx(:, conj_carriers) = conj(complex_matrix);

    % Plot OFDM Carriers on IFFT bins
    if fig == 1
        figure(1);
        stem(1:ifft_size, abs(spectrum_tx(2, :)), 'b*-');
        grid on;
        axis([0 ifft_size -0.5 1.5]);
        ylabel('Magnitude of PSK Data');
        xlabel('IFFT Bin');
        title('OFDM Carriers on designated IFFT bins');

        figure(2);
        plot(1:ifft_size, (180/pi) * angle(spectrum_tx(2, 1:ifft_size)), 'go');
        hold on;
        grid on;
        stem(carriers, (180/pi) * angle(spectrum_tx(2, carriers)), 'b*-');
        stem(conj_carriers, (180/pi) * angle(spectrum_tx(2, conj_carriers)), 'b*-');
        axis([0 ifft_size -200 +200]);
        ylabel('Phase (degree)');
        xlabel('IFFT Bin');
        title('Phases of the OFDM modulated Data');
    end

    % Obtain time wave from spectra using IFFT
    signal_tx = real(ifft(spectrum_tx'))';

    % Plot time signals
    if fig == 1
        % OFDM Time Signal (one symbol period in one carrier)
        limt = 1.1 * max(abs(reshape(signal_tx', 1, size(signal_tx, 1) * size(signal_tx, 2))));
        figure(3);
        plot(1:ifft_size, signal_tx(2, :));
        grid on;
        axis([0 ifft_size -limt limt]);
        ylabel('Amplitude');
        xlabel('Time');
        title('OFDM Time Signal (one symbol period in one carrier)');

        % Samples of OFDM Time Signals over one symbol period
        figure(4);
        colors = ['b', 'g', 'r', 'c', 'm', 'y'];
        for k = 1:min(length(colors), (carrier_symb_count - 1))
            plot(1:ifft_size, signal_tx(k + 1, :), colors(k));
            hold on;
        end
        grid on;
        axis([0 ifft_size -limt limt]);
        ylabel('Amplitude');
        xlabel('Time');
        title('Samples of OFDM Time Signals over one symbol period');
    end

    % Add a periodic guard time
    end_symb = size(signal_tx, 2);  % end of a symbol period without guard
    signal_tx = [signal_tx(:, (end_symb - guard_time + 1):end_symb) signal_tx];

    % Parallel to serial
    signal_tx = signal_tx';  % MATLAB's reshape goes along with columns
    signal_tx = reshape(signal_tx, 1, size(signal_tx, 1) * size(signal_tx, 2));
end

File: ofdm-matlab-py/ofdm_modulate.py
import numpy as np
from scipy.io import loadmat
import matplotlib.pyplot as plt

def ofdm_modulate(data_tx, ifft_size, carriers, conj_carriers, carrier_count, symb_size, guard_time, fig):
    carrier_symb_count = int(np.ceil(len(data_tx) / carrier_count))

    if len(data_tx) / carrier_count != carrier_symb_count:
        padding = np.zeros(carrier_symb_count * carrier_count)
        padding[:len(data_tx)] = data_tx
        data_tx = padding
    
    data_tx_matrix = np.reshape(data_tx, (carrier_symb_count, carrier_count)).T

    carrier_symb_count = data_tx_matrix.shape[1] + 1
    # generate random list
    # diff_ref = np.round(np.random.rand(carrier_count) * (2**symb_size) + 0.5).reshape(1009, 1)
    diff_ref = loadmat('../matlab-sim/diff_ref.mat')
    diff_ref = diff_ref['diff_ref'][0].reshape(1009, 1)
    
    data_tx_matrix = np.hstack((diff_ref, data_tx_matrix))
    

    for k in range(1, data_tx_matrix.shape[1]):
        data_tx_matrix[:, k] = np.remainder(data_tx_matrix[:, k] + data_tx_matrix[:, k-1], 2**symb_size)    

    ones_array = np.ones_like(data_tx_matrix)

    X = np.round(np.cos(data_tx_matrix * (2* np.pi / 2**symb_size)) * ones_array)
    Y = np.round(np.sin(data_tx_matrix * (2* np.pi / 2**symb_size)) * ones_array)

    complex_matrix = X + 1j * Y

    spectrum_tx = np.zeros((ifft_size, carrier_symb_count), dtype=complex)
    spectrum_tx[carriers[0]:carriers[-1]+1, :] = complex_matrix  
    spectrum_tx[conj_carriers[-1]-2:conj_carriers[0]-1, :] = np.conjugate(complex_matrix)
    # safe until here

    # if fig == 1:
    #     plt.figure(1)
    #     plt.stem(np.arange(1, ifft_size + 1), np.abs(spectrum_tx[1, :]), 'b*-')
    #     plt.grid(True)
    #     plt.axis([0, ifft_size, -0.5, 1.5])
    #     plt.ylabel('Magnitude of PSK Data')
    #     plt.xlabel('IFFT Bin')
    #     plt.title('OFDM Carriers on deiignated IFFT bins')

    #     plt.figure(2)
    #     plt.plot(np.arange(1, ifft_size+1), (180/np.pi) * np.angle(spectrum_tx[1, 0:ifft_size]), 'go')
    #     plt.hold(True)
    #     plt.grid(True)
    #     plt.stem(carriers, (180/np.pi) * np.angle(spectrum_tx[1, carriers], 'b*-'))
    #     plt.stem(conj_carriers, (180/np.pi) * np.angle(spectrum_tx[1, conj_carriers]), 'b*-')
    #     plt.axis([0, ifft_size, -200, 200])
    #     plt.ylabel('Phase (degree)')
    #     plt.xlabel('IFFT Bin')
    #     plt.title('Phases of the OFDM modulated data')

    # signal_tx = np.real(np.fft.ifft(spectrum_tx.T).T)

    # if fig == 1:
    #     limt = 1.1 * np.max(np.abs(np.reshape(signal_tx, (1, signal_tx.size))))
    #     plt.figure(3)
    #     plt.plot(np.arange(1, ifft_size+1), signal_tx[1, :])
    #     plt.grid(True)
    #     plt.axis([0, ifft_size, -limt, limt])
    #     plt.ylabel('Amplitude')
    #     plt.xlabel('Time')
    #     plt.title('OFDM Time Signal (one symbol period in one carrier)')

    #     plt.figure(4)
    #     colors = ['b', 'g', 'r', 'c', 'm', 'y']
    #     for k in range(1, min(len(colors), carrier_symb_count -1)):
    #         plt.plot(np.arange(1, ifft_size+1), signal_tx[k+1, :], colors[k])
    #         plt.hold(True)
    #     plt.gird(True)
    #     plt.axis([0, ifft_size, -limt, limt])
    #     plt.ylabel('Amplitude')
    #     plt.xlabel('Time')
    #     plt.title('Samples of OFDM Time Signals over one symbol period')

    # # Add a periodic guard time
    # end_symb = signal_tx.shape[1]
    # signal_tx = np.concatenate((signal_tx[:, end_symb - guard_time:end_symb], signal_tx), axis=1)

    # signal_tx = np.reshape(signal_tx.T, (1, signal_tx.size))

    # return signal_tx
